### ISSUE TITLE
Exceeded the default maximum gRPC message size

### ISSUE THREAD

**Issue by pupo-matheus** [istio/istio#52815]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

We recently migrated from Istio version 1.17 to 1.22.1. For nearly 10 days, everything worked properly until suddenly, Istio stopped working and we began throwing the error below when navigating through some pages

```
upstream connect error or disconnect/reset before headers. reset reason: connection timeout
```

Initially, we restarted the Gateway and Istiod services, temporarily alleviating the problem. However, the issue soon reoccurred. Upon analyzing the data plane logs, we found the following messages

```
warn    xdsproxy        upstream terminated with unexpected error rpc error: code = ResourceExhausted desc = grpc: received message larger than max (4243942 vs. 4000000)        id=111

2024-08-19T11:21:21.395733Z     warning envoy config external/envoy/source/extensions/config_subscription/grpc/grpc_stream.h:155        DeltaAggregatedResources gRPC config stream to xds-grpc closed: 2, send error for type url type.googleapis.com/envoy.config.listener.v3.Listener: EOF     thread=20
```

We reverted our environment to the previous version, and everything stabilized. After that, we investigated the root cause and discovered that the error was likely due to the gRPC upstream (Istio’s pilot/control plane) rejecting data plane messages that exceeded the configured maximum size. We could make the components communicate properly by setting the ISTIO_GPRC_MAXRECVMSGSIZE environment variable to a value above the default.

We would like to understand whether this is a bug or expected behavior, and also learn how we can monitor this so that we know when we are approaching the mentioned limits.

### Version

```Text
❯ istioctl version
client version: 1.23.0
control plane version: 1.22.1
data plane version: 1.22.1 (2 proxies)

❯ kubectl version
Client Version: v1.31.0
Kustomize Version: v5.4.2
Server Version: v1.30.2-eks-db838b0

❯ helm version
version.BuildInfo{Version:"v3.15.4", GitCommit:"fa9efb07d9d8debbb4306d72af76a383895aa8c4", GitTreeState:"clean", GoVersion:"go1.22.6"}
```

### Affected product area

- [X] Ambient
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [X] Networking
- [X] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [X] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

**Comment by howardjohn** [comment]
How can the request payload be so large when there is only 2 proxies? Is it a giant cluster and just a small mesh?

**Comment by pupo84** [comment]
It is indeed a cluster with a high volume of deployments. We currently have more than 7,000 virtual services.

Something I haven’t mentioned yet, but that might be relevant, is that we use Istio together with Knative in our architecture

**Comment by pupo-matheus** [comment]
It is indeed a cluster with a high volume of deployments. We currently have more than 7,000 virtual services.

Something I haven’t mentioned yet, but that might be relevant, is that we use Istio together with Knative in our architecture

**Comment by howardjohn** [comment]
Ah got it thanks! Just saw the 2 proxies so was confused thinking it was small. In that case the right solution for you is to configure that env var and just need to see if we have metrics for it. I think we might only have response size not request size

**Comment by pupo-matheus** [comment]
Thanks for the quick reply.

Could setting the environment variable and increasing the request size lead to any side effects?

Regarding metrics, it would be helpful to have the necessary data to understand when we might encounter any issues and to take action to mitigate them. I would be grateful for this information.

**Comment by howardjohn** [comment]
Well right now it's hard failing, so raising it is really the only option unless you can reduce the amount of configuration. The risk is really that you obviously have a massive config and it's only getting more massive, which may lead to performance issues

**Comment by pupo-matheus** [comment]
I understand that reducing the size of the cluster is the best option, but I would still like to find a way to monitor for problematic behavior to avoid being caught off guard by the same issues that prompted this concern

**Comment by howardjohn** [comment]
I don't think there are currently these metrics

**Comment by hzxuzhonghu** [comment]
don't we set grpc.MaxCallRecvMsgSize to max uint32?

**Comment by howardjohn** [comment]
only for the things reading xds responses, this is xds request

**Comment by hzxuzhonghu** [comment]
so how could request be so large? Even with 7000 vs, it can hardly reach 4MB

**Comment by howardjohn** [comment]
yeah that surprised me too. `istioctl x internal-debug adsz` would maybe
give a clue

**Comment by pupo-matheus** [comment]
The command doesn't even finish due to a timeout error

**Comment by pupo-matheus** [comment]
I was able to execute the command. I’ve masked some sensitive information, but this should give you an idea of what’s in the cluster

[istio-config.tgz](https://github.com/user-attachments/files/16949868/istio-config.tgz)

**Comment by howardjohn** [comment]
```
$ cat istio-config.json | jq '.clients[0].watches["type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment"][]' -r | wc -l
23322
```

Yikes. 1.8mb just from the names (note: this is the size of the debug output, the gRPC representation is not 1:1, and there is more than jsut the names), so it.

@pupo-matheus one thing I notice is there are tons of duplicates:

```
1502 outbound|9091||sfj-xxxxxxx-xxxxxxx-00001.namespace-private.svc.cluster.local
```

for instance.

Are those because you masked them? or are they really duplicate somehow (that would be super weird). Can check with

```
 cat istio-config.json | jq '.clients[0].watches["type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment"][]' -r | sort | uniq -c | sort -n
```

**Comment by pupo-matheus** [comment]
Yes, it's due to the masking

**Comment by howardjohn** [comment]
One thing you could do is set `PILOT_FILTER_GATEWAY_CLUSTER_CONFIG=true`. Part of the problem is each Knative service generates 2 services with 8 ports between them. You probably only use 1, so the size is 8x what it needs to be. The env var should filter out things that are not used, which at least in my reproduction goes from 8 ports -> 2.

That being said, its intentionally set to false due to some downtime scenarios in some cases: https://github.com/istio/istio/issues/29131.

**Comment by pupo-matheus** [comment]
Is it safe to use this approach? I’ve seen some reports of issues with it.
